internal/rootcoord/quota_center.go

```diff
@@ -844,118 +1248,214 @@ func (q *QuotaCenter) getCollectionLimitProperties(collection int64) map[string]
 }
 
 // checkDiskQuota checks if disk quota exceeded.
-func (q *QuotaCenter) checkDiskQuota() {
+func (q *QuotaCenter) checkDiskQuota() error {
 	q.diskMu.Lock()
 	defer q.diskMu.Unlock()
 	if !Params.QuotaConfig.DiskProtectionEnabled.GetAsBool() {
-		return
+		return nil
 	}
 	if q.dataCoordMetrics == nil {
-		return
+		return nil
 	}
-	collections := typeutil.NewUniqueSet()
+
+	denyCluster := false
 	totalDiskQuota := Params.QuotaConfig.DiskQuota.GetAsFloat()
+	total := q.dataCoordMetrics.TotalBinlogSize
+	if float64(total) >= totalDiskQuota {
+		denyCluster = true
+	}
+
+	collectionDiskQuota := Params.QuotaConfig.DiskQuotaPerCollection.GetAsFloat()
+	dbSizeInfo := make(map[int64]int64)
+	collections := make([]int64, 0)
 	for collection, binlogSize := range q.dataCoordMetrics.CollectionBinlogSize {
 		collectionProps := q.getCollectionLimitProperties(collection)
-		colDiskQuota := getCollectionRateLimitConfig(collectionProps, common.CollectionDiskQuotaKey)
+		colDiskQuota := getRateLimitConfig(collectionProps, common.CollectionDiskQuotaKey, collectionDiskQuota)
```

Comment:
1. Checking db size,  force deny writing them.
2. Check collection size in the next step and skip collection that in force deny abs.
3. checking partition size and skipping partitions that in force deny collections.